File: app/src/test/java/com/babestudios/companyinfouk/ui/charges/ChargesPresenterTest.java
package com.babestudios.companyinfouk.ui.charges;

import com.babestudios.companyinfouk.data.DataManager;
import com.babestudios.companyinfouk.data.model.charges.Charges;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.junit.MockitoJUnitRunner;

import io.reactivex.Observable;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class ChargesPresenterTest {

	private ChargesPresenter chargesPresenter;

	@Before
	public void setUp() {
		chargesPresenter = new ChargesPresenter(mock(DataManager.class));
		ChargesActivityView view = mock(ChargesActivityView.class);
		when(view.getCompanyNumber()).thenReturn("0");
		when(chargesPresenter.getDataManager().getCharges("0", "0")).thenReturn(Observable.just(new Charges()));
		chargesPresenter.create();
		chargesPresenter.attachView(view);

	}

	@Test
	public void whenGetInsolvency_thenDataManagerGetInsolvencyIsCalled() {
		chargesPresenter.getCharges();
		verify(chargesPresenter.getDataManager(), times(2)).getCharges("0", "0");
	}

	@Test
	public void whenLoadMoreCharges_thenDataManagerLoadMoreChargesIsCalled() {
		chargesPresenter.loadMoreCharges(0);
		verify(chargesPresenter.getDataManager(), times(2)).getCharges("0", "0");
	}
}

File: app/src/test/java/com/babestudios/companyinfouk/ui/charges/ChargesPresenterTest.kt
package com.babestudios.companyinfouk.ui.charges

import com.babestudios.companyinfouk.CompaniesHouseApplication
import com.babestudios.companyinfouk.DaggerTestApplicationComponent
import com.babestudios.companyinfouk.TestApplicationModule
import com.babestudios.companyinfouk.data.CompaniesRepositoryContract
import com.babestudios.companyinfouk.data.model.charges.Charges
import com.nhaarman.mockitokotlin2.whenever
import io.reactivex.Completable
import io.reactivex.Single
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import org.mockito.junit.MockitoJUnitRunner

@RunWith(MockitoJUnitRunner::class)
class ChargesPresenterTest {

	private lateinit var chargesPresenter: ChargesPresenter

	@Before
	fun setUp() {
		val testApplicationComponent = DaggerTestApplicationComponent.builder()
				.testApplicationModule(TestApplicationModule(CompaniesHouseApplication()))
				.build()
		chargesPresenter = testApplicationComponent.chargesPresenter()
		whenever(chargesPresenter.companiesRepository.fetchCharges("123", "0")).thenReturn(Single.just(Charges()))
		val chargesViewModel = ChargesViewModel()
		chargesViewModel.state.value.companyNumber = "123"
		chargesPresenter.setViewModel(chargesViewModel, Completable.fromCallable { false })
	}

	@Test
	fun whenGetInsolvency_thenDataManagerGetInsolvencyIsCalled() {
		chargesPresenter.fetchCharges("123")
		verify<CompaniesRepositoryContract>(chargesPresenter.companiesRepository, times(2)).fetchCharges("123", "0")
	}

	@Test
	fun whenLoadMoreCharges_thenDataManagerLoadMoreChargesIsCalled() {
		chargesPresenter.loadMoreCharges(0)
		verify<CompaniesRepositoryContract>(chargesPresenter.companiesRepository, times(2)).fetchCharges("123", "0")
	}
}